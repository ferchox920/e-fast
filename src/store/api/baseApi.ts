// src/store/api/baseApi.ts
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import type { RootState } from '@/store';

export const baseApi = createApi({
  reducerPath: 'api',
  // Añade 'UserList' a los tagTypes
  tagTypes: ['User', 'UserList'], // <-- Modifica esta línea
  baseQuery: fetchBaseQuery({
    baseUrl:
      process.env.NEXT_PUBLIC_API_BASE_URL?.replace(/\/$/, '') ?? 'http://127.0.0.1:8000/api/v1',
    prepareHeaders: (headers, { getState }) => {
      const token = (getState() as RootState).user.token;
      if (token) headers.set('authorization', `Bearer ${token}`);
      return headers;
    },
  }),
  endpoints: () => ({}),
});
