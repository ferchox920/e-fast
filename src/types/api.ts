// src/types/api.ts
import type { UUID } from './common';

export interface ApiError {
  code: string | number;
  message: string;
  details?: unknown;
  trace_id?: UUID | string;
}

export interface ApiResponse<T> {
  success: boolean;
  data: T;
  message?: string;
  error?: ApiError;
}

/** Paginación genérica */
export interface Paginated<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

/** Parámetros genéricos de listado */
export interface PaginationParams {
  page?: number;
  page_size?: number;
}

export type SortOrder = 'asc' | 'desc';

export interface ListParamsBase extends PaginationParams {
  search?: string;
  sort_by?: string;
  sort_order?: SortOrder;
}
