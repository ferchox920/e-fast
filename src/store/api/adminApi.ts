// src/store/api/adminApi.ts
import { baseApi } from './baseApi';
import type { UserRead, UserCreate } from '@/types/user';
import type { Paginated, PaginationParams } from '@/types/api';

// Antes: interface ListUsersParams extends PaginationParams {}
// Ahora: type (no dispara no-empty-object-type)
type ListUsersParams = PaginationParams;

type CreateUserAdminBody = Partial<UserCreate> &
  Pick<UserCreate, 'email' | 'password'> & {
    is_superuser?: boolean;
  };

interface SetUserRoleParams {
  userId: string;
  makeAdmin: boolean;
}

interface SetUserActiveParams {
  userId: string;
  active: boolean;
}

export const adminApi = baseApi.injectEndpoints({
  endpoints: (build) => ({
    listUsers: build.query<Paginated<UserRead>, ListUsersParams>({
      query: ({ page = 1, page_size = 50 }) => ({
        url: '/admin/users',
        params: { page, page_size },
      }),
      providesTags: (result) =>
        result
          ? [
              ...result.items.map(({ id }) => ({ type: 'User' as const, id })),
              { type: 'UserList' as const, id: 'LIST' },
            ]
          : [{ type: 'UserList' as const, id: 'LIST' }],
    }),

    createUserAdmin: build.mutation<UserRead, CreateUserAdminBody>({
      query: ({ is_superuser = false, ...body }) => ({
        url: '/admin/users',
        method: 'POST',
        body: { ...body, is_superuser },
      }),
      invalidatesTags: [{ type: 'UserList', id: 'LIST' }],
    }),

    setUserRole: build.mutation<UserRead, SetUserRoleParams>({
      query: ({ userId, makeAdmin }) => ({
        url: `/admin/users/${userId}/role`,
        method: 'PATCH',
        body: { make_admin: makeAdmin },
      }),
      invalidatesTags: (result, error, { userId }) => [
        { type: 'User', id: userId },
        { type: 'UserList', id: 'LIST' },
      ],
    }),

    setUserActiveStatus: build.mutation<UserRead, SetUserActiveParams>({
      query: ({ userId, active }) => ({
        url: `/admin/users/${userId}/active`,
        method: 'PATCH',
        body: { active },
      }),
      invalidatesTags: (result, error, { userId }) => [
        { type: 'User', id: userId },
        { type: 'UserList', id: 'LIST' },
      ],
    }),
  }),
});

export const {
  useListUsersQuery,
  useLazyListUsersQuery,
  useCreateUserAdminMutation,
  useSetUserRoleMutation,
  useSetUserActiveStatusMutation,
} = adminApi;
