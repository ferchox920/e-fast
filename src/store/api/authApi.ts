import { baseApi } from './baseApi';
import type { UserRead } from '@/types/user';

// Tipos según tus endpoints de FastAPI
export type TokenPair = {
  access_token: string;
  refresh_token: string;
  token_type: 'bearer';
  expires_in: number;
  user: UserRead;
};

export type TokenRefresh = {
  access_token: string;
  token_type: 'bearer';
  expires_in: number;
};

type LoginBody = { email: string; password: string };
type RefreshBody = { refresh_token: string };
type VerifyEmailBody = { email: string };

export const authApi = baseApi.injectEndpoints({
  endpoints: (build) => ({
    // /auth/login usa OAuth2PasswordRequestForm (username + password) x-www-form-urlencoded
    // /auth/login (OAuth2PasswordRequestForm)
    login: build.mutation<TokenPair, { email: string; password: string }>({
      query: ({ email, password }) => ({
        url: '/auth/login',
        method: 'POST',
        body: new URLSearchParams({ username: email, password }),
      }),
      invalidatesTags: ['User'],
    }),

    // /auth/refresh { refresh_token }
    refresh: build.mutation<TokenRefresh, RefreshBody>({
      query: (body) => ({
        url: '/auth/refresh',
        method: 'POST',
        body,
      }),
    }),

    // /auth/verify/request { email } → 204
    requestVerifyEmail: build.mutation<void, VerifyEmailBody>({
      query: (body) => ({
        url: '/auth/verify/request',
        method: 'POST',
        body,
      }),
    }),
  }),
  overrideExisting: false,
});

export const { useLoginMutation, useRefreshMutation, useRequestVerifyEmailMutation } = authApi;
