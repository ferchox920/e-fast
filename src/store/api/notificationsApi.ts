import { baseApi } from './baseApi';
import type { NotificationRead, NotificationUpdate } from '@/types/notifications';

export interface ListNotificationsParams {
  limit?: number;
  offset?: number;
}

export interface ListNotificationsResponse {
  items: NotificationRead[];
  total: number;
  limit: number;
  offset: number;
}

const sortByCreatedAtDesc = (a: NotificationRead, b: NotificationRead) => {
  if (a.created_at === b.created_at) return 0;
  return a.created_at < b.created_at ? 1 : -1;
};

export const notificationsApi = baseApi.injectEndpoints({
  endpoints: (build) => ({
    listNotifications: build.query<ListNotificationsResponse, ListNotificationsParams | void>({
      query: ({ limit = 50, offset = 0 } = {}) => ({
        url: '/notifications',
        params: { limit, offset },
      }),
      transformResponse: (response: ListNotificationsResponse) => ({
        ...response,
        items: [...(response.items ?? [])].sort(sortByCreatedAtDesc),
      }),
      providesTags: (result) => [
        ...(result?.items?.map(({ id }) => ({ type: 'Notification' as const, id })) ?? []),
        { type: 'NotificationList' as const },
      ],
    }),
    updateNotification: build.mutation<
      NotificationRead,
      { id: NotificationRead['id'] } & NotificationUpdate
    >({
      query: ({ id, ...body }) => ({
        url: `/notifications/${id}`,
        method: 'PATCH',
        body,
      }),
      invalidatesTags: (result, error, { id }) => [
        { type: 'Notification' as const, id },
        { type: 'NotificationList' as const },
      ],
    }),
  }),
});

export const {
  useListNotificationsQuery,
  useLazyListNotificationsQuery,
  useUpdateNotificationMutation,
} = notificationsApi;
