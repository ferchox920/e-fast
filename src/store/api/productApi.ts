// src/store/api/productApi.ts
import { baseApi } from './baseApi';
import type {
  PaginatedProducts,
  ProductCreate,
  ProductImageCreate,
  ProductImageRead,
  ProductListParams,
  ProductRead,
  ProductUpdate,
  ProductVariantCreate,
  ProductVariantRead,
  ProductVariantUpdate,
} from '@/types/product';

type UpdateProductArgs = {
  productId: string;
  body: ProductUpdate;
};

type CreateVariantArgs = {
  productId: string;
  body: ProductVariantCreate;
};

type UpdateVariantArgs = {
  variantId: string;
  body: ProductVariantUpdate;
};

type DeleteVariantArgs = {
  variantId: string;
  productId?: string;
};

type AddImageArgs = {
  productId: string;
  body: ProductImageCreate;
};

type SetPrimaryImageArgs = {
  productId: string;
  imageId: string;
};

export const productApi = baseApi.injectEndpoints({
  endpoints: (build) => ({
    listProducts: build.query<PaginatedProducts, ProductListParams | undefined>({
      query: (params) => ({
        url: '/products',
        params: params ? params : undefined,
      }),
      providesTags: (result) =>
        result
          ? [
              ...result.items.map(({ id, slug }) => ({ type: 'Product' as const, id: id ?? slug })),
              { type: 'ProductList' as const, id: 'LIST' },
            ]
          : [{ type: 'ProductList' as const, id: 'LIST' }],
    }),

    getProduct: build.query<ProductRead, string>({
      query: (slug) => ({
        url: `/products/${slug}`,
      }),
      providesTags: (result, error, slug) => [{ type: 'Product' as const, id: result?.id ?? slug }],
    }),

    createProduct: build.mutation<ProductRead, ProductCreate>({
      query: (body) => ({
        url: '/products',
        method: 'POST',
        body,
      }),
      invalidatesTags: [{ type: 'ProductList', id: 'LIST' }],
    }),

    updateProduct: build.mutation<ProductRead, UpdateProductArgs>({
      query: ({ productId, body }) => ({
        url: `/products/${productId}`,
        method: 'PUT',
        body,
      }),
      invalidatesTags: (result, error, { productId }) => [
        { type: 'Product', id: result?.id ?? productId },
        { type: 'ProductList', id: 'LIST' },
      ],
    }),

    listVariants: build.query<ProductVariantRead[], string>({
      query: (productId) => ({
        url: `/products/${productId}/variants`,
      }),
      providesTags: (result, error, productId) => {
        const variantTags =
          result?.map((variant) => ({ type: 'ProductVariant' as const, id: variant.id })) ?? [];
        return [...variantTags, { type: 'Product', id: productId }];
      },
    }),

    createVariant: build.mutation<ProductVariantRead, CreateVariantArgs>({
      query: ({ productId, body }) => ({
        url: `/products/${productId}/variants`,
        method: 'POST',
        body,
      }),
      invalidatesTags: (result, error, { productId }) => [
        { type: 'Product', id: productId },
        { type: 'ProductList', id: 'LIST' },
      ],
    }),

    updateVariant: build.mutation<ProductVariantRead, UpdateVariantArgs>({
      query: ({ variantId, body }) => ({
        url: `/variants/${variantId}`,
        method: 'PUT',
        body,
      }),
      invalidatesTags: (result, error, { variantId }) => [
        { type: 'ProductVariant' as const, id: variantId },
        ...(result?.product_id ? [{ type: 'Product' as const, id: result.product_id }] : []),
      ],
    }),

    deleteVariant: build.mutation<void, DeleteVariantArgs>({
      query: ({ variantId }) => ({
        url: `/variants/${variantId}`,
        method: 'DELETE',
      }),
      invalidatesTags: (result, error, { variantId, productId }) => {
        const baseTags = [
          { type: 'ProductVariant' as const, id: variantId },
          { type: 'ProductList' as const, id: 'LIST' },
        ];
        return productId ? [...baseTags, { type: 'Product' as const, id: productId }] : baseTags;
      },
    }),

    addImage: build.mutation<ProductImageRead, AddImageArgs>({
      query: ({ productId, body }) => ({
        url: `/products/${productId}/images`,
        method: 'POST',
        body,
      }),
      invalidatesTags: (result, error, { productId }) => [
        { type: 'Product' as const, id: result?.product_id ?? productId },
        { type: 'ProductList', id: 'LIST' },
        ...(result?.id ? [{ type: 'ProductImage' as const, id: result.id }] : []),
      ],
    }),

    setPrimaryImage: build.mutation<ProductRead, SetPrimaryImageArgs>({
      query: ({ productId, imageId }) => ({
        url: `/products/${productId}/images/${imageId}/primary`,
        method: 'POST',
      }),
      invalidatesTags: (result, error, { productId }) => [
        { type: 'Product', id: result?.id ?? productId },
        { type: 'ProductList', id: 'LIST' },
      ],
    }),
  }),
});

export const {
  useListProductsQuery,
  useLazyListProductsQuery,
  useGetProductQuery,
  useLazyGetProductQuery,
  useCreateProductMutation,
  useUpdateProductMutation,
  useListVariantsQuery,
  useCreateVariantMutation,
  useUpdateVariantMutation,
  useDeleteVariantMutation,
  useAddImageMutation,
  useSetPrimaryImageMutation,
} = productApi;
