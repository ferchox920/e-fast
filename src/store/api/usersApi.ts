import { baseApi } from './baseApi';
import type { UserCreate, UserRead, UserUpdate } from '@/types/user';

export type RegisterBody = Pick<UserCreate, 'email' | 'password' | 'full_name'> &
  Partial<UserCreate>;

export const usersApi = baseApi.injectEndpoints({
  endpoints: (build) => ({
    /**
     * POST /users -> Crea usuario
     */
    register: build.mutation<UserRead, RegisterBody>({
      query: (body) => ({
        url: '/users',
        method: 'POST',
        body,
      }),
      invalidatesTags: ['User'],
    }),

    /**
     * GET /users/me -> Perfil del usuario autenticado
     */
    me: build.query<UserRead, void>({
      query: () => ({ url: '/users/me' }),
      providesTags: ['User'],
    }),

    /**
     * PUT /users/me -> Actualiza el perfil del usuario autenticado
     */
    updateMe: build.mutation<UserRead, UserUpdate>({
      query: (body) => ({
        url: '/users/me',
        method: 'PUT',
        body,
      }),
      invalidatesTags: ['User'],
    }),
  }),
});

export const {
  useRegisterMutation,
  useMeQuery,
  useLazyMeQuery, // ðŸ‘ˆ export del hook lazy
  useUpdateMeMutation,
} = usersApi;
