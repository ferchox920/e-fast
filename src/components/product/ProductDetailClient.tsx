'use client';

import { useMemo } from 'react';
import { useGetProductQuery } from '@/store/api/productApi';
import ProductGallery from './ProductGallery';
import ProductGallerySkeleton from './ProductGallerySkeleton';

interface ProductDetailClientProps {
  slug: string;
}

export function ProductDetailClient({ slug }: ProductDetailClientProps) {
  const { data, isLoading, isFetching, isError, error, refetch } = useGetProductQuery(slug, {
    skip: !slug,
    refetchOnMountOrArgChange: true,
  });

  const isPending = isLoading || isFetching;

  const anomalyMessages = useMemo(() => {
    if (!data?.images?.length) return [] as string[];
    const primaryCount = data.images.filter((image) => image.is_primary).length;
    const messages: string[] = [];
    if (primaryCount === 0) {
      messages.push(
        'No hay imagen principal definida. Se mostrará la primera imagen ordenada como fallback.',
      );
    }
    if (primaryCount > 1) {
      messages.push(
        'Se recibieron varias imágenes principales. Se mostrará la primera según sort_order.',
      );
    }
    return messages;
  }, [data?.images]);

  const galleryContent = useMemo(() => {
    if (isPending) return <ProductGallerySkeleton />;
    if (!data) return null;
    return <ProductGallery images={data.images} />;
  }, [data, isPending]);

  return (
    <section className="mx-auto flex w-full max-w-5xl flex-col gap-6 p-6">
      <header className="space-y-2">
        <p className="text-sm text-neutral-500">Producto</p>
        <h1 className="text-3xl font-semibold tracking-tight text-neutral-900">
          {data?.title ?? 'Cargando producto...'}
        </h1>
      </header>

      {isError && (
        <div className="space-y-3 rounded-md border border-red-200 bg-red-50 p-4 text-sm text-red-700">
          <p>
            No pudimos cargar el producto.{` `}
            {error && 'status' in error ? `Código: ${String(error.status)}` : null}
          </p>
          <button
            type="button"
            onClick={() => void refetch()}
            className="inline-flex items-center justify-center gap-2 rounded-md border border-red-600 px-3 py-1.5 text-xs font-semibold text-red-700 transition hover:bg-red-100 focus:outline-none focus:ring-2 focus:ring-red-400"
          >
            Reintentar
          </button>
        </div>
      )}

      {anomalyMessages.length ? (
        <div className="space-y-1 rounded-md border border-amber-200 bg-amber-50 p-3 text-sm text-amber-700">
          {anomalyMessages.map((message) => (
            <p key={message}>{message}</p>
          ))}
        </div>
      ) : null}

      {galleryContent ?? (
        <div className="flex aspect-4/3 items-center justify-center rounded-lg border border-dashed border-neutral-200 bg-neutral-50 text-sm text-neutral-500">
          Este producto no tiene imágenes disponibles.
        </div>
      )}

      {data && (
        <article className="space-y-4">
          {data.description && (
            <p className="text-base leading-relaxed text-neutral-700">{data.description}</p>
          )}
          <dl className="grid grid-cols-1 gap-4 sm:grid-cols-2">
            <div className="rounded-lg border border-neutral-200 bg-white p-4">
              <dt className="text-xs uppercase tracking-wide text-neutral-500">Precio</dt>
              <dd className="text-xl font-semibold text-neutral-900">
                {new Intl.NumberFormat('es-ES', {
                  style: 'currency',
                  currency: data.currency ?? 'ARS',
                }).format(data.price)}
              </dd>
            </div>
            <div className="rounded-lg border border-neutral-200 bg-white p-4">
              <dt className="text-xs uppercase tracking-wide text-neutral-500">Estado</dt>
              <dd className="text-sm text-neutral-800">
                {data.active ? 'Disponible' : 'Inactivo'}
              </dd>
            </div>
          </dl>
        </article>
      )}
    </section>
  );
}

export default ProductDetailClient;
