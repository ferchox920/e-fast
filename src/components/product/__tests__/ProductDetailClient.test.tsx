import { render } from '@testing-library/react';
import type { ProductImageRead, ProductRead } from '@/types/product';

jest.mock('@/store/api/productApi', () => ({
  useGetProductQuery: jest.fn(),
}));

import { useGetProductQuery } from '@/store/api/productApi';
import ProductDetailClient from '../ProductDetailClient';

const mockedUseGetProductQuery = useGetProductQuery as jest.Mock;

const baseProduct: ProductRead = {
  id: 'prod-1',
  slug: 'my-product',
  title: 'Producto Demo',
  description: 'Descripcion demo',
  price: 199,
  currency: 'USD',
  active: true,
  images: [],
  variants: [],
  created_at: null,
  updated_at: null,
};

const createImage = (id: string, overrides: Partial<ProductImageRead> = {}): ProductImageRead => ({
  id,
  product_id: baseProduct.id,
  url: `https://example.com/${id}.jpg`,
  alt_text: `Imagen ${id}`,
  is_primary: false,
  sort_order: 1,
  ...overrides,
});

describe('ProductDetailClient snapshots', () => {
  afterEach(() => {
    mockedUseGetProductQuery.mockReset();
  });

  it('renders loading state snapshot', () => {
    mockedUseGetProductQuery.mockReturnValue({
      data: null,
      isLoading: true,
      isFetching: true,
      isError: false,
      error: null,
      refetch: jest.fn(),
    });

    const { asFragment } = render(<ProductDetailClient slug="my-product" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders error state with retry button', () => {
    mockedUseGetProductQuery.mockReturnValue({
      data: null,
      isLoading: false,
      isFetching: false,
      isError: true,
      error: { status: 500 },
      refetch: jest.fn(),
    });

    const { asFragment, getByRole } = render(<ProductDetailClient slug="my-product" />);
    expect(getByRole('button', { name: /Reintentar/i })).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows anomaly warning when multiple primary images', () => {
    mockedUseGetProductQuery.mockReturnValue({
      data: {
        ...baseProduct,
        images: [
          createImage('1', { is_primary: true, sort_order: 1 }),
          createImage('2', { is_primary: true, sort_order: 2 }),
        ],
      },
      isLoading: false,
      isFetching: false,
      isError: false,
      error: null,
      refetch: jest.fn(),
    });

    const { getByText, asFragment } = render(<ProductDetailClient slug="my-product" />);
    expect(getByText(/Se recibieron varias im√°genes principales/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
