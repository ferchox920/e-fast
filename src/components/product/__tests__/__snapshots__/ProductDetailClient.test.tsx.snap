// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductDetailClient snapshots renders error state with retry button 1`] = `
<DocumentFragment>
  <section
    class="mx-auto flex w-full max-w-5xl flex-col gap-6 p-6"
  >
    <header
      class="space-y-2"
    >
      <p
        class="text-sm text-neutral-500"
      >
        Producto
      </p>
      <h1
        class="text-3xl font-semibold tracking-tight text-neutral-900"
      >
        Cargando producto...
      </h1>
    </header>
    <div
      class="space-y-3 rounded-md border border-red-200 bg-red-50 p-4 text-sm text-red-700"
    >
      <p>
        No pudimos cargar el producto. Código: 500
      </p>
      <button
        class="inline-flex items-center justify-center gap-2 rounded-md border border-red-600 px-3 py-1.5 text-xs font-semibold text-red-700 transition hover:bg-red-100 focus:outline-none focus:ring-2 focus:ring-red-400"
        type="button"
      >
        Reintentar
      </button>
    </div>
    <div
      class="flex aspect-4/3 items-center justify-center rounded-lg border border-dashed border-neutral-200 bg-neutral-50 text-sm text-neutral-500"
    >
      Este producto no tiene imágenes disponibles.
    </div>
  </section>
</DocumentFragment>
`;

exports[`ProductDetailClient snapshots renders loading state snapshot 1`] = `
<DocumentFragment>
  <section
    class="mx-auto flex w-full max-w-5xl flex-col gap-6 p-6"
  >
    <header
      class="space-y-2"
    >
      <p
        class="text-sm text-neutral-500"
      >
        Producto
      </p>
      <h1
        class="text-3xl font-semibold tracking-tight text-neutral-900"
      >
        Cargando producto...
      </h1>
    </header>
    <div
      class="space-y-3"
      data-testid="product-gallery-skeleton"
    >
      <div
        aria-hidden="true"
        class="aspect-[4/3] w-full animate-pulse rounded-lg bg-neutral-200"
      />
      <div
        class="flex gap-2"
      >
        <div
          aria-hidden="true"
          class="h-20 w-20 animate-pulse rounded-md bg-neutral-200"
        />
        <div
          aria-hidden="true"
          class="h-20 w-20 animate-pulse rounded-md bg-neutral-200"
        />
        <div
          aria-hidden="true"
          class="h-20 w-20 animate-pulse rounded-md bg-neutral-200"
        />
        <div
          aria-hidden="true"
          class="h-20 w-20 animate-pulse rounded-md bg-neutral-200"
        />
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`ProductDetailClient snapshots shows anomaly warning when multiple primary images 1`] = `
<DocumentFragment>
  <section
    class="mx-auto flex w-full max-w-5xl flex-col gap-6 p-6"
  >
    <header
      class="space-y-2"
    >
      <p
        class="text-sm text-neutral-500"
      >
        Producto
      </p>
      <h1
        class="text-3xl font-semibold tracking-tight text-neutral-900"
      >
        Producto Demo
      </h1>
    </header>
    <div
      class="space-y-1 rounded-md border border-amber-200 bg-amber-50 p-3 text-sm text-amber-700"
    >
      <p>
        Se recibieron varias imágenes principales. Se mostrará la primera según sort_order.
      </p>
    </div>
    <div
      class="space-y-3"
    >
      <figure
        class="relative aspect-4/3 overflow-hidden rounded-lg border border-neutral-200 bg-neutral-50"
      >
        <img
          alt="Imagen 1"
          class="object-cover"
          data-nimg="fill"
          data-testid="product-gallery-main-image"
          decoding="async"
          id="product-gallery-main-image"
          loading="lazy"
          sizes="(min-width:1024px) 640px, (min-width:640px) 560px, 100vw"
          src="/_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
        <figcaption
          class="sr-only"
        >
          Imagen 1
        </figcaption>
      </figure>
      <div
        aria-label="Imagenes del producto"
        class="flex gap-2 overflow-x-auto"
        data-testid="product-gallery-thumbnails"
        role="listbox"
      >
        <button
          aria-controls="product-gallery-main-image"
          aria-selected="true"
          class="relative size-20 shrink-0 overflow-hidden rounded-md border transition focus:outline-none focus:ring-2 focus:ring-neutral-600 ring-2 ring-neutral-900 border-neutral-900"
          data-gallery-thumb="true"
          role="option"
          type="button"
        >
          <span
            class="sr-only"
          >
            Imagen seleccionada: Imagen 1
          </span>
          <img
            alt="Imagen 1"
            class="object-cover"
            data-nimg="1"
            decoding="async"
            height="80"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2F1.jpg&w=256&q=75 2x"
            style="color: transparent;"
            width="80"
          />
        </button>
        <button
          aria-controls="product-gallery-main-image"
          aria-selected="false"
          class="relative size-20 shrink-0 overflow-hidden rounded-md border transition focus:outline-none focus:ring-2 focus:ring-neutral-600 border-neutral-200"
          data-gallery-thumb="true"
          role="option"
          type="button"
        >
          <span
            class="sr-only"
          >
            Seleccionar imagen: Imagen 2
          </span>
          <img
            alt="Imagen 2"
            class="object-cover"
            data-nimg="1"
            decoding="async"
            height="80"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fexample.com%2F2.jpg&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fexample.com%2F2.jpg&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2F2.jpg&w=256&q=75 2x"
            style="color: transparent;"
            width="80"
          />
        </button>
      </div>
    </div>
    <article
      class="space-y-4"
    >
      <p
        class="text-base leading-relaxed text-neutral-700"
      >
        Descripcion demo
      </p>
      <dl
        class="grid grid-cols-1 gap-4 sm:grid-cols-2"
      >
        <div
          class="rounded-lg border border-neutral-200 bg-white p-4"
        >
          <dt
            class="text-xs uppercase tracking-wide text-neutral-500"
          >
            Precio
          </dt>
          <dd
            class="text-xl font-semibold text-neutral-900"
          >
            199,00 US$
          </dd>
        </div>
        <div
          class="rounded-lg border border-neutral-200 bg-white p-4"
        >
          <dt
            class="text-xs uppercase tracking-wide text-neutral-500"
          >
            Estado
          </dt>
          <dd
            class="text-sm text-neutral-800"
          >
            Disponible
          </dd>
        </div>
      </dl>
    </article>
  </section>
</DocumentFragment>
`;
